### ISSUE TITLE
Volunteers needed: CMake makefiles need updating

### ISSUE THREAD

**Issue by Dead2** [zlib-ng/zlib-ng#8]
CMake makefiles have not been updated since the fork from zlib.

The makefiles needs to get cleaned up and updated, support for old compilers can be removed if it brings a benefit to the project.

**Comment by mtl1979** [comment]
Building with cmake using CMakeLists.txt is now working again... 
Next step is to figure out if we want to remove something unnecessary or add something that was added to configure after forking from official zlib.
I tested that cmake does not panic when building out of source tree, so that part is in pretty good state.
There is some code in configure that refers to files meant exclusively for cmake, but should be in my opinion isolated as cmake is actually meant to replace configure, not make.

**Comment by nemequ** [comment]
Continuing the discussion from #40…

> The current proper way of getting these flags set is running ./configure, since that makes the proper makefile and sets up arch support and such. If you can run that before compiling, then that would currently be the best way to auto-detect flags. More complete CMake support is coming at some point.

I'll try to take some time in the next week or two to update CMake support to match the configure script, but I have a couple questions first:

I suggested changing the architecture of the CMake stuff in an earlier comment on that issue so that tests are done in a separate file which is then included by CMakeLists.txt to allow sharing more easily between zlib-ng and software which is embedding a copy of it.  Are you okay with that idea?

Do you want to keep the two systems functioning side-by-side, or are you trying to switch to CMake?  I have a project called [configure-cmake](https://github.com/nemequ/configure-cmake) which could be used to provide an autotools-style configure script which calls CMake behind the scenes, it could be a good way to consolidate the build system while maintaining a comfortable environment for people who are used to configure scripts.  The only real down side (IMHO) is that it currently requires bash (not csh, dash, tcsh, etc.)… AFAIK pretty much everyone other than FreeBSD ships bash by default (it might not be the default shell, but it is at least installed).

**Comment by mtl1979** [comment]
cmake builds still have some compatibility/performance issues, so I think we should keep the configure script separate for now. There is already work on modularizing CMakeLists.txt and I see no reason why it could not be done now rather than later as we have been rewriting a lot of it already.

**Comment by Dead2** [comment]
Please have a look at RJVB and mtl1979's forks, they have made a lot of progress on CMake. Possibly it is even ready for a pull request after cleaning up the commits, but I haven't had a chance to look into it myself yet.

I am unsure what splitting the CMake test stuff would entail, but it might very well be a good idea. I am not familiar enough with CMake myself to see the pros/cons and any problems it might cause. Regarding tests, have a look at daxtens fork. He was working on a new testing framework that I would very much like to have in place soon.

I do think we need to keep configure separate for now, splitting that out can be a possible long-term goal. Also I seem to remember somebody saying that CMake is a no-go on some platform we want to support, cant remember the details at the moment though.

**Comment by mtl1979** [comment]
The remaining issues with both configure and CMake were mostly to work around instruction set detecting when compiler is a cross-compiler, because it is pointless to execute test code to check if it crashes or not on target arch.

**Comment by nemequ** [comment]
Okay, I think I'll wait until you have a chance to look at the forks before I do anything.  I don't want to generate a bunch of conflicts.

> Also I seem to remember somebody saying that CMake is a no-go on some platform we want to support, cant remember the details at the moment though.

If you do remember I would be very interested.  AFAIK CMake should be more portable than pretty much anything else.

> The remaining issues with both configure and CMake were mostly to work around instruction set detecting when compiler is a cross-compiler, because it is pointless to execute test code to check if it crashes or not on target arch.

Yeah, I noticed that too.  For example, the X86_QUICK_STRATEGY stuff depends on SSE2, so it should be determined at runtime.  I wasn't sure if that was just enabling a runtime check or hardcoding an SSE2 dependency.

It's not just an issue for cross-compiling.  For example, most Linux distros expect to compile packages which will work on all machines of a specific architecture, not just those with some specific version of SSE, NEON, AVX, etc.

Anyways, I would prefer not to have to duplicate lots of stuff in Squash, so once the main repository is in better shape let me know and I'll take a look at what still needs to be done to easily configure an optimal build for software embedding zlib-ng.

**Comment by mtl1979** [comment]
There is two issues... Too much run-time checking makes the code slower but more universal, on the other hand disabling run-time checks is faster but can make the code crash on certain processors that support one of the newer instruction sets but not all the ones before it. This is mostly issue with all the different variations of SSE4.x and AVX.

**Comment by nemequ** [comment]
There are ways around that.  If your run-time checks are taking up too much time the right thing to do (IMHO) is limit sharing of _compiled_ code, not increase your dependence on compile-time checks for things which should be tested at run-time.  Macros and static/inline functions can be help ensure source code is shared.  You may even have to compile or include a source file multiple times using different compiler flags or function attributes, using the preprocessor to change the symbol names to generate versions with different optimizations.

**Comment by mtl1979** [comment]
I know there is solution for every problem, but in this case it comes down to fact that not all the developers have machines that have SSE4.2 so we can't test the code that it actually works as expected. We can do test builds that other people can test, but if even one machine fails to compile or run the build, it's a lost cause.

**Comment by nemequ** [comment]
That's a good point.  I wonder what versions of SSE Travis supports.  It should be pretty easy to automatically test SSE versions <= that version in CI.  If it's too low, maybe drone.io or one of the others runs on newer machines…

**Comment by grooverdan** [comment]
https://github.com/madler/zlib/pull/125 shows an example of using a IFUNC to dynamically select optimized code at load time. It fails back to using a static function pointer when this GNU loader extension isn't supported.

**Comment by mtl1979** [comment]
@grooverdan That patch can be reduced in size by avoiding use of uInt and other user types we have already discarded in favor of standard types.

**Comment by grooverdan** [comment]
totally agree, the patch needs rework in those ways and using the arch directories to be considered for zlib-ng. I'll see what I can done time permitting. I was just mentioning the IFUNC method of dynamic load time one time determining of hardware capabilities as a way to see if anyone objected or wants to see it done differently.

**Comment by mtl1979** [comment]
@grooverdan We only need to make sure it doesn't break for any weird reason when the code is compiled under Visual C++ or any other non-gcc compiler...

**Comment by nemequ** [comment]
It might be nice to have something a bit more generic and centralized, so you have one initialization function where you detect the CPU features and choose the optimized versions.  Something like this (untested, but you should get the idea):

``` C
struct {
  void (* doSomething)(void);
  void (* doSomethingElse)(void)
} zlibNgFuncTable = {
  doSomethingInit,
  doSomethingElseInit
};

static once_flag zlibNgFuncTableOnce = ONCE_INIT;

static void doSomethingInStraightC () {
  /* ... */
}

static void doSomethingWithSSE3 () {
  /* ... */
}

static void doSomethingWithSSE4 () {
  /* ... */
}

/* Similar functions for doSomethingElse */

void zlibNgInit () {
  unsigned int sse_version = useCpuIdToGetSupportedVersion();

  if (see_version >= SSE_VERSION_4_0) {
    zlibNgFuncTable.doSomething = doSomethingWithSSE4;
  } else if (sse_version > SSE_VERSION_3_0) {
    zlibNgFuncTable.doSomething = doSomethingWithSSE3;
  } else {
    zlibNgFuncTable.doSomething = doSomethingWithStraightC;
  }

  /* Something similar for doSomethingElse, preferably integrated into the same tests. */
}

static void doSomethingInit () {
  call_once (&zlibNgFuncTableOnce, zlibNgInit);
  zlibNgFuncTable.doSoemthing ();
}

static void doSomethingElseInit () {
  call_once (&zlibNgFuncTableOnce, zlibNgInit);
  zlibNgFuncTable.doSoemthing ();
}
```

I used C11 threads stuff, and didn't even look to see what zlib-ng's coding style looked like, but that's just superficial stuff.

Then, to invoke the function just call the appropriate callback in the table.  Of course you could also get rid of the struct and just have the callbacks at the top level.  Obviously I like the struct, but it doesn't really matter.

**Comment by mtl1979** [comment]
@nemequ What you described was similar what we had planned but haven't had time to implement yet... The outer layer needs to be as compiler-agnostic as possible, but inner layer can be optimized for different compilers.

**Comment by nemequ** [comment]
@mtl1979 I might be willing to put something like that together.  I haven't really been interested in touching the code because last time I looked at the project I had the impression there was a lot of unmerged stuff waiting around and I didn't want to step on anyone's toes, but if that's no longer the case…

Does anyone know what the MSVC equivalent of GCC's [function-specific option pragmas](https://gcc.gnu.org/onlinedocs/gcc/Function-Specific-Option-Pragmas.html#Function-Specific-Option-Pragmas) (i.e., how to target a specific processor extension on a per-function basis) is?

**Comment by mtl1979** [comment]
@nemequ As far as I know, there is only one unmerged patch remaining... We have merged a lot of the useful code recently...
